File: src/lib/jev/fp/span.ts
import type { Color } from "../colors.ts";
import { err, ok, type Result } from "./result.ts";
import type { GlueError, System } from "./cospan.ts";

/** A ← S → B. Nadir S is the shared variable. */
export type Span = {
  readonly left: System;
  readonly right: System;
  readonly nadir: Color;
};

export type Pullback = {
  readonly color: Color;
  readonly name: string;
};

export function spanOf(left: System, right: System, color: Color): Result<Span, GlueError> {
  const hasL = left.ports.some((p) => p.color === color);
  const hasR = right.ports.some((p) => p.color === color);
  if (!hasL || !hasR) return err({ kind: "color-mismatch", color });
  return ok({ left, right, nadir: color });
}

/** Discrete pullback — twin of Rust `pullback_ports`. */
export function pullback(s: Span): Pullback {
  return { color: s.nadir, name: `${s.left.name} ∩ ${s.right.name}` };
}

export function share(left: System, right: System, color: Color): Result<Pullback, GlueError> {
  const s = spanOf(left, right, color);
  return s.ok ? ok(pullback(s.value)) : s;
}

File: crates/jev-kernel/src/span.rs
//! Spans of colored interfaces.
//!
//! A span is A ← S → B. The nadir S is the shared variable.
//! Composition of variable-sharing systems is the pullback:
//! two views of the same color collapse to one state.

use crate::cospan::GlueError;

/// Discrete pullback: the shared color, if both feet carry it.
pub fn pullback_ports(left: &[u8], right: &[u8], color: u8) -> Result<Vec<u8>, GlueError> {
    if left.contains(&color) && right.contains(&color) {
        Ok(vec![color])
    } else {
        Err(GlueError::ColorMismatch)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn share_entity() {
        assert_eq!(pullback_ports(&[0, 3], &[0, 1], 0).unwrap(), vec![0]);
    }
}
